File: python/tempmail_wrapper/providers/ten_minute_mail.py
"""10minutemail.net provider implementation (HTML scraping public mailbox)."""

from __future__ import annotations

import re
from datetime import datetime

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


def decode_cf_email(hex_str: str) -> str:
    if len(hex_str) < 4:
        return ""
    try:
        k = int(hex_str[0:2], 16)
        return "".join(
            chr(int(hex_str[i:i+2], 16) ^ k)
            for i in range(2, len(hex_str), 2)
        )
    except ValueError:
        return ""


def strip_html(html: str) -> str:
    return re.sub(r"<[^>]+>", "", html).strip()


class TenMinuteMailProvider(TempMailProvider):
    """Provider for 10minutemail.net — HTML scraping public mailbox."""

    BASE_URL = "https://10minutemail.net"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=True)
        self._email: str | None = None

    @property
    def name(self) -> str:
        return "10minutemail"

    def generate_email(self) -> str:
        resp = self._http.get(f"{self.BASE_URL}/", timeout=30)
        if resp.status_code != 200:
            raise TempMailError(f"Failed to get email: {resp.status_code}", self.name)
        html = resp.text
        match = re.search(r'id="fe_text"[^>]*value="([^"]+)"', html, re.IGNORECASE)
        if not match:
            raise TempMailError("No address in response", self.name)
        self._email = match.group(1).strip()
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        resp = self._http.get(f"{self.BASE_URL}/mailbox.ajax.php", timeout=30)
        if resp.status_code != 200:
            raise TempMailError(f"Inbox error: {resp.status_code}", self.name)
        html = resp.text

        rows = re.findall(r"<tr[^>]*>(.*?)</tr>", html, re.DOTALL | re.IGNORECASE)
        if len(rows) <= 1:
            return []

        messages = []
        # Skip header row
        for row in rows[1:]:
            cells = re.findall(r"<td[^>]*>(.*?)</td>", row, re.DOTALL | re.IGNORECASE)
            if len(cells) < 3:
                continue

            # Extract sender
            sender = ""
            cf_match = re.search(r'data-cfemail="([^"]+)"', cells[0], re.IGNORECASE)
            if cf_match:
                sender = decode_cf_email(cf_match.group(1))
            else:
                sender = strip_html(cells[0])

            subject = strip_html(cells[1])

            # Extract Date
            date_str = ""
            date_match = re.search(r'title="([^"]+)"', cells[2], re.IGNORECASE)
            if date_match:
                date_str = date_match.group(1)
            else:
                date_str = strip_html(cells[2])

            try:
                if "utc" not in date_str.lower():
                    date_str += " UTC"
                # e.g., 2026-07-02 15:45:54 UTC
                date = datetime.strptime(date_str, "%Y-%m-%d %H:%M:%S %Z")
            except Exception:
                date = datetime.now()

            # Message ID
            mid_match = re.search(r"mid=([^'&\"\s>]+)", row, re.IGNORECASE)
            if not mid_match:
                continue
            msg_id = mid_match.group(1)

            messages.append(
                Message(
                    id=msg_id,
                    sender=sender,
                    subject=subject,
                    date=date,
                )
            )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        # composite message_id handling (e.g. welcome:email)
        mid = message_id.split(":")[0] if ":" in message_id else message_id

        resp = self._http.get(f"{self.BASE_URL}/readmail.html?mid={mid}", timeout=30)
        if resp.status_code != 200:
            raise NotFoundError(f"Cannot read message: {resp.status_code}", self.name)
        html = resp.text

        body_match = re.search(
            r'class="mailinhtml"[^>]*>(.*?)<div[^>]*style="clear:both;"',
            html,
            re.DOTALL | re.IGNORECASE,
        )
        if not body_match:
            raise NotFoundError(f"Message {message_id} not found", self.name)

        body_html = body_match.group(1).strip()

        # Decode Cloudflare obfuscated email links inside body
        body_html = re.sub(
            r'<(?:a|span)[^>]*class="__cf_email__"[^>]*data-cfemail="([^"]+)"[^>]*>.*?</(?:a|span)>',
            lambda m: decode_cf_email(m.group(1)),
            body_html,
            flags=re.IGNORECASE,
        )
        body_html = re.sub(
            r'href="/cdn-cgi/l/email-protection#([^"]+)"',
            lambda m: f'href="mailto:{decode_cf_email(m.group(1))}"',
            body_html,
            flags=re.IGNORECASE,
        )

        body_text = strip_html(body_html)

        # Subject
        subject_match = re.search(
            r'<div class="mail_header">.*?<h2[^>]*>(.*?)</h2>',
            html,
            re.DOTALL | re.IGNORECASE,
        )
        subject = strip_html(subject_match.group(1)) if subject_match else ""

        # Sender
        from_match = re.search(
            r'<span class="mail_from">(.*?)</span>',
            html,
            re.DOTALL | re.IGNORECASE,
        )
        sender = ""
        if from_match:
            cf_from = re.search(
                r'data-cfemail="([^"]+)"', from_match.group(1), re.IGNORECASE
            )
            sender = (
                decode_cf_email(cf_from.group(1))
                if cf_from
                else strip_html(from_match.group(1))
            )

        return MessageDetail(
            id=mid,
            sender=sender,
            subject=subject,
            date=datetime.now(),
            body_text=body_text,
            body_html=body_html,
            attachments=[],
        )

    def delete_email(self, email: str) -> bool:
        self._email = None
        return True

File: rust/src/providers/ten_minute_mail.rs
use async_trait::async_trait;
use regex::Regex;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const BASE_URL: &str = "https://10minutemail.net";

/// Provider for 10minutemail.net — HTML scraping public mailbox.
#[derive(Debug)]
pub struct TenMinuteMail {
    client: HttpClient,
    email: Arc<Mutex<Option<String>>>,
}

impl TenMinuteMail {
    pub fn new(client: HttpClient) -> Self {
        Self {
            client,
            email: Arc::new(Mutex::new(None)),
        }
    }
}

// Decode Cloudflare obfuscated email address
fn decode_cf_email(hex: &str) -> String {
    if hex.len() < 4 {
        return String::new();
    }
    let k = match u8::from_str_radix(&hex[0..2], 16) {
        Ok(val) => val,
        Err(_) => return String::new(),
    };
    let mut decoded = String::new();
    for i in (2..hex.len()).step_by(2) {
        if i + 2 <= hex.len() {
            if let Ok(val) = u8::from_str_radix(&hex[i..i+2], 16) {
                decoded.push((val ^ k) as char);
            }
        }
    }
    decoded
}

fn strip_html(html: &str) -> String {
    let re = Regex::new(r"<[^>]+>").unwrap();
    re.replace_all(html, "").trim().to_string()
}

#[async_trait]
impl TempMailProvider for TenMinuteMail {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        let html = self.client.get(&format!("{}/", BASE_URL)).await?;
        let re = Regex::new(r#"(?i)id="fe_text"[^>]*value="([^"]+)""#).unwrap();
        let cap = re.captures(&html)
            .ok_or_else(|| TempMailError::Api("No address in response".into()))?;
        let email = cap.get(1).unwrap().as_str().trim().to_string();
        let mut lock = self.email.lock().await;
        *lock = Some(email.clone());
        Ok(email)
    }

    async fn get_inbox(&self, _email: &str) -> Result<Vec<Message>, TempMailError> {
        let html = self.client.get(&format!("{}/mailbox.ajax.php", BASE_URL)).await?;
        
        let re_row = Regex::new(r"(?is)<tr[^>]*>(.*?)</tr>").unwrap();
        let re_cell = Regex::new(r"(?is)<td[^>]*>(.*?)</td>").unwrap();
        let re_cf = Regex::new(r#"(?i)data-cfemail="([^"]+)""#).unwrap();
        let re_title = Regex::new(r#"(?i)title="([^"]+)""#).unwrap();
        let re_mid = Regex::new(r"(?i)mid=([^'&'\s>]+)").unwrap();

        let rows: Vec<_> = re_row.captures_iter(&html).collect();
        if rows.len() <= 1 {
            return Ok(vec![]);
        }

        let mut messages = vec![];
        // Skip header row
        for cap in rows.iter().skip(1) {
            let row_html = cap.get(1).unwrap().as_str();
            let cells: Vec<_> = re_cell.captures_iter(row_html).collect();
            if cells.len() < 3 {
                continue;
            }

            // Sender
            let cell_0 = cells[0].get(1).unwrap().as_str();
            let sender = if let Some(cf_cap) = re_cf.captures(cell_0) {
                decode_cf_email(cf_cap.get(1).unwrap().as_str())
            } else {
                strip_html(cell_0)
            };

            let subject = strip_html(cells[1].get(1).unwrap().as_str());

            // Date
            let cell_2 = cells[2].get(1).unwrap().as_str();
            let date_str = if let Some(title_cap) = re_title.captures(cell_2) {
                title_cap.get(1).unwrap().as_str().to_string()
            } else {
                strip_html(cell_2)
            };

            let mut date_str_utc = date_str;
            if !date_str_utc.to_lowercase().contains("utc") {
                date_str_utc.push_str(" UTC");
            }
            
            let date = chrono::NaiveDateTime::parse_from_str(&date_str_utc, "%Y-%m-%d %H:%M:%S UTC")
                .map(|ndt| chrono::DateTime::<chrono::Utc>::from_naive_utc_and_offset(ndt, chrono::Utc))
                .unwrap_or_else(|_| chrono::Utc::now());

            // Mid
            let id = if let Some(mid_cap) = re_mid.captures(row_html) {
                mid_cap.get(1).unwrap().as_str().to_string()
            } else {
                continue;
            };

            messages.push(Message {
                id,
                sender,
                subject,
                date,
                ..Default::default()
            });
        }

        Ok(messages)
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        let mid = if let Some(idx) = message_id.find(':') {
            &message_id[..idx]
        } else {
            message_id
        };

        let html = self.client.get(&format!("{}/readmail.html?mid={}", BASE_URL, mid)).await?;

        let re_body = Regex::new(r#"(?is)class="mailinhtml"[^>]*>(.*?)<div[^>]*style="clear:both;""#).unwrap();
        let cap_body = re_body.captures(&html)
            .ok_or(TempMailError::NotFound)?;
        let mut body_html = cap_body.get(1).unwrap().as_str().trim().to_string();

        // Decode CF email obfuscation in body
        let re_cf_link = Regex::new(r#"(?i)<(?:a|span)[^>]*class="__cf_email__"[^>]*data-cfemail="([^"]+)"[^>]*>.*?</(?:a|span)>"#).unwrap();
        body_html = re_cf_link.replace_all(&body_html, |caps: &regex::Captures| {
            decode_cf_email(caps.get(1).unwrap().as_str())
        }).to_string();

        let re_cf_prot = Regex::new(r#"(?i)href="/cdn-cgi/l/email-protection#([^"]+)""#).unwrap();
        body_html = re_cf_prot.replace_all(&body_html, |caps: &regex::Captures| {
            format!("href=\"mailto:{}\"", decode_cf_email(caps.get(1).unwrap().as_str()))
        }).to_string();

        let body_text = strip_html(&body_html);

        // Subject
        let re_sub = Regex::new(r#"(?is)<div class="mail_header">.*?<h2[^>]*>(.*?)</h2>"#).unwrap();
        let subject = if let Some(sub_cap) = re_sub.captures(&html) {
            strip_html(sub_cap.get(1).unwrap().as_str())
        } else {
            String::new()
        };

        // Sender
        let re_from = Regex::new(r#"(?is)<span class="mail_from">(.*?)</span>"#).unwrap();
        let sender = if let Some(from_cap) = re_from.captures(&html) {
            let from_html = from_cap.get(1).unwrap().as_str();
            let re_cf_from = Regex::new(r#"(?i)data-cfemail="([^"]+)""#).unwrap();
            if let Some(cf_from_cap) = re_cf_from.captures(from_html) {
                decode_cf_email(cf_from_cap.get(1).unwrap().as_str())
            } else {
                strip_html(from_html)
            }
        } else {
            String::new()
        };

        Ok(MessageDetail {
            message: Message {
                id: mid.to_string(),
                sender,
                subject,
                date: chrono::Utc::now(),
                ..Default::default()
            },
            body_text: if body_text.is_empty() { None } else { Some(body_text) },
            body_html: if body_html.is_empty() { None } else { Some(body_html) },
            attachments: vec![],
            ..Default::default()
        })
    }

    async fn delete_email(&self, _email: &str) -> Result<bool, TempMailError> {
        let mut e = self.email.lock().await;
        *e = None;
        Ok(true)
    }
}
